### ISSUE TITLE
[Question] Day template child note templates?

### ISSUE THREAD

**Issue by PAK90** [zadam/trilium#1062]
I want to have a Day template that has three child notes; morning, afternoon and evening, each with their own specific label attributes. Creating a Day template with sub-notes is easy (by linking them into the Day template note), but when I create a Day note with the default 'Today' button, the linked notes are clones of the morning/afternoon/evening notes, which means they're not very useful (i.e. when I click on the button, I essentially want 4 notes created; the root Day note and the three child ones). Above that, I'd like their titles to include some date info (e.g. `30 - Saturday - Morning`) automatically.

Is there a way to do this purely with the UI, or is it necessary to modify the 'Today' button script? I can imagine something where, for example, the Morning template has a label that the API searches for, copies (not clones) that label, and makes today's copy appear under the Day note. Is that the only viable approach?

**Comment by zadam** [comment]
Hello, I'm a bit confused by you mentioning clones and labels, but to distill the basic idea - day note is sort of a copy/duplicate of day template, but currently only the template itself is duplicated and not its children/subtree.

I agree that this would make sense, however this is not implemented (yet). For now I'd recommend modifying the Today button script - from that you can create the subnotes, make their titles more dynamic (including the date e.g.) etc.

**Comment by PAK90** [comment]
Ok, I figured it'd be a case of modifying the script. This gets into another question... how do I copy a note using the API? I have a 'morning template' (just a normal note with some text and a label: `@morningDayRoot`), so in the script I find it with:

    const todayMorningNote = await api.searchForNote('@morningDayRoot');

This returns the right note (checked via console.log), but doing `api.activateNote` does nothing; I think I need to copy it (somehow) then set that copy as a child of the current todayNote. There's backend API for setting relations, but not for copying; is this possible?

**Comment by zadam** [comment]
There isn't any API specific to copying, only setting as you mention. So you need to implement something like:

```
newNote.title = tplNote.title;
await newNote.setContent(await tplNote.getContent());

for (const tplAttr of await tpl.getAttributes()) {
	await new Attribute({
		noteId: newNote.noteId,
		type: tplAttr.type,
		name: tplAttr.name,
		value: tplAttr.value,
		isInheritable: tplAttr.isInheritable
	}).save();
}
```

**Comment by PAK90** [comment]
I'm having some issues getting the server code to work, so I had two questions; first one; if I do a `console.log` in server code, where does that get output? And second one; how do I read params from `runOnBackend`? Here's what I have:

        const todayNote = await api.getTodayNote();
        console.log(todayNote.noteId);
        
        await api.waitUntilSynced();
        api.activateNote(todayNote.noteId);
        
        // run backend to create morning note copy
        const returnedMorningNote = await api.runOnBackend(async ([parentId]) => {
            
            const todayMorningNote = await api.getNoteWithLabel('morningDayRoot');
            const morningContent = await todayMorningNote.getContent();
            console.log(parentId); // where does this output to??
            
            const morningNote = await api.createNewNote(
                parentId,
                'morning note',
                morningContent,
                'text',
            );
            for (const tplAttr of await todayMorningNote.getAttributes()) {
                await new Attribute({
                    noteId: morningNote.noteId,
                    type: tplAttr.type,
                    name: tplAttr.name,
                    value: tplAttr.value,
                    isInheritable: tplAttr.isInheritable
                }).save();
            }
            return morningNote;
        }, [todayNote.noteId]);
        
        console.log(returnedMorningNote);
        api.activateNote(returnedMorningNote.noteId);

What I see in client console is it giving the todayNote's ID, then an error: `Parent note undefined not found.`, which I assume means the `parentId` I'm passing in is undefined? But... it shouldn't be. I tried the non-array param too (i.e. `(parentId) => {}` instead of `([parentId]) => {}`) but that didn't work either.

**Comment by zadam** [comment]
> if I do a `console.log` in server code, where does that get output?

In the standard output of the backend process. I recommend using `api.log("text")` in the backend API which will log it into logs (`trilium-data/log/...`) and you can see them also in Top-left menu -> Show backend log.

Here's the fixed code:

```javascript
const todayNote = await api.getTodayNote();
console.log(todayNote.noteId);

await api.waitUntilSynced();
api.activateNote(todayNote.noteId);

// run backend to create morning note copy
const returnedMorningNote = await api.runOnBackend(async (parentId) => {

    const todayMorningNote = await api.getNoteWithLabel('morningDayRoot');
    const morningContent = await todayMorningNote.getContent();
    api.log("PARENT_ID: " + parentId); // where does this output to??

    const {note: morningNote} = await api.createTextNote(
        parentId,
        'morning note',
        morningContent
    );
    for (const tplAttr of await todayMorningNote.getOwnedAttributes()) {
        await morningNote.addAttribute(tplAttr.type, tplAttr.name, tplAttr.value, tplAttr.isInheritable);
    }
    return morningNote;
}, [todayNote.noteId]);

console.log(returnedMorningNote);
api.activateNote(returnedMorningNote.noteId);
```
Few things there:

* there was extra array destructuring of parentId parameter (`[parentId]` instead of just `parentId`)
* createNewNote was using an incorrect contract. Could you mention where did you get this? If it's in some docs I should fix it ...
* I gave you a bad advice regarding the attributes:
  * need to use `getOwnedAttributes()` instead of `getAttributes()` since the latter would recreate again inherited attributes which would be wrong
  * directly creating entities of Attribute is not supported, so it's necessary to use note's API

**Comment by PAK90** [comment]
Thank you for the fixed code, it works rather well! In regards to `createNewNote`, in the backend script docs, it says it takes `CreateNewNoteParams`, so that's what I tried to do.

One question I did have; the morning note that ends up being created has both the `labelDefinition` and `label` versions of the `labelDefinition` attributes that I'm defining in the morning template; is there a way to make the day morning note only have the labels, and not the labelDefinitions, since they're redundant? Something like the Book template in the example docs.

**Comment by zadam** [comment]
You can just filter based on the attribute type, so like this:

```javascript
for (const tplAttr of await todayMorningNote.getOwnedAttributes()) {
    if (tplAttr.type === 'label) {	
        await morningNote.addAttribute(tplAttr.type, tplAttr.name, tplAttr.value, tplAttr.isInheritable);
    }
}
```

**Comment by PAK90** [comment]
What ended up working for me was setting a relation called 'template' for the copied morningNote, pointing to the template morningNote, and then I didn't have to do any messing around with copying attributes at all. Is that the intended functionality?

**Comment by zadam** [comment]
Well, depends. When a note X has a template relation to the note Y, then X inherits all the Y's attributes. For some purposes it's OK ...

**Comment by zadam** [comment]
This has been implemented in stable branch and will be released in 0.45.5 soon.
